### PR TITLE
Fix Eslint & Add typescript Eslint & add some rules

### PR DESCRIPTION
**Pull request by Gusted**
Code Quality to the win. I will add some more rules later but this is mostly to fix up the workflow of eslint to play nice with Typescript.

### CODE REVIEW

File: src/data.ts [hunk 1]
```diff
@@ -43,6 +43,6 @@ export const setupSession = (set: Setter) => {
   })
 }
 
-export const authenticateUser = (provider: Provider) => client.auth.signIn({ provider })
+export const authenticateUser = async (provider: Provider) => client.auth.signIn({ provider })
```
Comment (bjornstar) [inline src/data.ts:46]: what's putting the async here?

Comment (Gusted) [reply src/data.ts:46]: `@typescript-eslint/promise-function-async` -> https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/promise-function-async.md

TL;DR if a function returns a Promise the `() => Promise<Result>` should be `async () => Result`

File: .eslintrc.json [hunk 2]
```diff
@@ -27,11 +24,41 @@
         "react/display-name": 0,
         "react/prop-types": 0,
         "react/react-in-jsx-scope": 0,
-        "semi": ["error", "never"]
+        "semi": ["error", "never"],
+        "no-debugger": "error",
+        "eol-last": ["error", "always"]
```
Comment (bjornstar) [inline .eslintrc.json:29]: 👍

File: package.json [hunk 3]
```diff
@@ -43,10 +45,10 @@
     "dev": "vite",
     "build": "tsc && vite build",
     "serve": "vite preview",
-    "eslint": "eslint --cache src",
-    "eslint-fix": "eslint --fix --cache src",
-    "prettier": "prettier --list-different '**/*.{js,jsx,md}'",
-    "prettier-fix": "prettier --write '**/*.{js,jsx,md}'",
+    "eslint": "eslint --cache \"src/**/*.ts\" \"src/**/*.tsx\" \"src/**/*.js\" \"src/**/*.jsx\"",
```
Comment (bjornstar) [inline package.json:48]: Can we leave the file extension overrides in `.eslintrc.json`?

File: package.json [hunk 4]
```diff
@@ -43,10 +45,10 @@
...
+    "eslint-fix": "eslint --fix --cache \"src/**/*.ts\" \"src/**/*.tsx\" \"src/**/*.js\" \"src/**/*.jsx\"",
+    "prettier": "prettier --list-different 'src/**/*.{js,jsx,md,ts,tsx}'",
```
Comment (bjornstar) [inline package.json:50]: adding src here seems like it will bite us in the future if we add another directory that needs to be prettified.

File: .eslintrc.json [hunk 5]
```diff
@@ -27,11 +24,41 @@
...
+    "overrides": [{
+        "files": ["src/**/*.tsx", "src/**/*.ts"],
+        "parser": "@typescript-eslint/parser",
+        "extends": "plugin:@typescript-eslint/recommended",
+        "plugins": ["@typescript-eslint"],
+        "parserOptions": {
+            "project": "tsconfig.json"
+        },
+
+        "rules": {
```
Comment (bjornstar) [inline .eslintrc.json:45]: Can we start with the defaults?

File: .eslintrc.json [hunk 6]
```diff
@@ -27,11 +24,41 @@
...
+        "eol-last": ["error", "always"]
     },
     "settings": {
         "react": {
             "version": "detect"
         }
-    }
+    },
+    "overrides": [{
+        "files": ["src/**/*.tsx", "src/**/*.ts"],
```
Comment (bjornstar) [inline .eslintrc.json:37]: Can we remove src here as well?

File: src/data.ts (hunk 1 above)
Comment (bjornstar) [reply src/data.ts:46]: I'm not a fan of adding extraneous `async` directives. It doesn't seem to be a default rule, can we get rid of it?

Comment (Gusted) [reply src/data.ts:46]: sure thing.

File: .eslintrc.json (hunk 5 above)
Comment (Gusted) [reply .eslintrc.json:45]: You mean the rules specified some lines above? These simply overwrite the previous one. So anything added their will be used here or otherwise specified in this section.

Comment (bjornstar) [reply .eslintrc.json:45]: I tried it out on my branch without custom rules for typescript and got reasonable results.

The warnings that it generated seemed like good things to fix but were not show stoppers.

File: .eslintrc.json [hunk 7]
```diff
@@ -27,11 +24,42 @@
...
+        "eol-last": ["error", "always"]
     },
     "settings": {
         "react": {
             "version": "detect"
         }
-    }
+    },
+    "overrides": [{
+        "files": ["**/*.tsx", "**/*.ts"],
```
Comment (bjornstar) [inline .eslintrc.json:37]:
```suggestion
        "files": ["*.tsx", "*.ts"],
```

File: .eslintrc.json [hunk 8]
```diff
@@ -8,9 +8,6 @@
         "plugin:react/recommended",
         "eslint:recommended"
     ],
-    "overrides": [{
-        "files": ["*.jsx", "*.js"]
```
Comment (bjornstar) [inline .eslintrc.json:12]: These need to get put back into the new overrides located down there VVV

Comment (Gusted) [reply .eslintrc.json:12]: 👍🏽

File: package.json [hunk 9]
```diff
@@ -43,10 +45,10 @@
     "dev": "vite",
     "build": "tsc && vite build",
     "serve": "vite preview",
-    "eslint": "eslint --cache src",
-    "eslint-fix": "eslint --fix --cache src",
-    "prettier": "prettier --list-different '**/*.{js,jsx,md}'",
-    "prettier-fix": "prettier --write '**/*.{js,jsx,md}'",
+    "eslint": "eslint --cache ./src",
+    "eslint-fix": "eslint --fix --cache ./src",
```
Comment (bjornstar) [inline package.json:49]: is this change necessary? Sorry for so many nits :cat:

Comment (Gusted) [reply package.json:49]: Nope - I'm used to `./` to specify local folder. UNIX? thingy.

File: .eslintrc.json (hunk 5 above)
Comment (Gusted) [reply .eslintrc.json:45]: Done

File: .eslintrc.json [hunk 10]
```diff
@@ -27,11 +24,45 @@
...
+
+            // "@typescript-eslint/array-type": ["error", {
+            //     "default": "array-simple"
+            // }],
+            // "@typescript-eslint/no-non-null-assertion": "off",
+            // "@typescript-eslint/no-redeclare": "error",
+            // "@typescript-eslint/explicit-module-boundary-types": "off",
+            // "@typescript-eslint/no-implied-eval": "error",
+            // "@typescript-eslint/switch-exhaustiveness-check": "error",
+            // "@typescript-eslint/consistent-type-imports": "error"
```
Comment (Gusted) [inline .eslintrc.json:62]: This is the one to ensure `import type [...]` @bjornstar

Comment (bjornstar) [reply .eslintrc.json:62]: Ah sweet, let's add just that one :)

Comment (Gusted) [reply .eslintrc.json]: Done

**Comment by bjornstar** [comment]
Thanks :)
